src/core/viz/expressions/ordering.js

```diff
@@ -4,8 +4,8 @@ import { checkInstance } from './utils';
 /**
  * Order ascending by a provided expression. NOTE: only works with `width()`.
  *
- * @param {carto.expressions.Base} by - must be `width()`
- * @return {carto.expressions.Base}
+ * @param {carto.expressions.width} by - must be `width()`
```

Comment:
I am using the function here to enable the JSDoc link (to the function)

Reply:
I see... but it seems really strange to me to point to a function when expecting a type. This case is not important because `order` is hidden, but for `property` a similar thing happened.

Reply:
Yes. Let's revisit this (width) later. With properties is different because a property is not a type itself: is something that has no type until it is evaluated for each feature. For example, the `add` expression allows expressions containing properties but of type Number. Another example: `cluster` aggregations only allow s.prop(''), no every expression, so I think in this case is more useful for the user/developer to be explicit and point to carto.expressions.prop.

Reply:
The point here is that we have dynamic types. This is not inheritance, no interfaces, just a `type` variable that in some cases is not static. The used Number, Category, Color, etc. represents just that, expressions (that could contain properties) so if I see `{Number|Color}` for the `add` I know that I cannot pass other expressions like Category, Date, etc. And the errors contain also that names of types. (I know we agree on this, but just to make it clear).

There is no issue with `viewport` because they can manage expressions as input parameters.
- `viewportAgg( Number ) -> Number`

BUT, we have a problem with `cluster` and `global` expressions.  They are really special functions because of the input and parameters. Cluster and Global aggregations only admit a property, nothing more. Property is not a type because its type is assigned when metadata is obtained (Number, Category). So we need to document that to represent this limitation. E.g. 

- `clusterAgg( carto.expressions.prop ) -> Number | Category
- `globalAgg( carto.expressions.prop ) -> Number

OR

- `clusterAgg( Number | Category ) -> Number | Category  BUT input can ONLY be properties (carto.expressions.prop), no other expressions.
- `globalAgg( Number ) -> Number  BUT input can ONLY be properties (carto.expressions.prop), no other expressions.

Reply:
Yes, that problem clearly exists.

Regarding the two options, the first one is in my opinion not very good for the reasons commented above, it is not technically correct since `carto.expressions.prop` is not a type, is a function.

The second option is ok to me.

However, I think there is a third option, and that is creating a `Property` type, that's not so unusual (see `Torque`). Of course, creating a `Property` type doesn't mean that `Property` couldn't be a subtype of `Number` and `Category`.

I don't agree with one thing you said at first, from my point of view, `Number` and `Category` are interfaces. Interfaces with no public API, but interfaces. Also, just because a function expects an interface of type X, and you pass a valid type that complies with the interface it doesn't mean that the function cannot do further validation and be more restrictive. For example, in general, you can divide any two numbers (type checking), but divide by zero and things will go wrong (runtime problem). I see the same problem here, our divide by zero scenario is the having a property which resolves to an unexpected type.

Reply:
I think the discussion about Property being a type (not sure about this one) and adding subtypes plus proper TypeScript interfaces, etc is a good one, but I think it is also a big one. Therefore it is out of the scope of this PR and the coming release.

So, if we are all >0.5 happy with the second version, let's do that and address/continue the discussion of types, subtypes in future development.

Reply:
Adding Typescript is clearly outside the scope of this. However, being correct about the types is not. I use Typescript as an example of hey, if Typescript wouldn't compile these types, that's because they are not correct. And before the PR Typescript would compile, of course, that's because we had less restrictive types in the documentation, but, that's exactly my point: we cannot put in the docs types that are so restrictive that using properties is not something valid according to our documented types.

Reply:
Just to clarify :)

In order to be consistent with the option 2, we must specify that the `s.prop` expression returns a `Number|Category`. Then, we have a "consistent" declaration for all the expressions using the basic types.

Reply:
Great!